components/make-static-previews.py

```diff
@@ -0,0 +1,145 @@
+    for compdir in components(basedir):
+        meta, _ = get_qmd_split(compdir / "index.qmd")
+        if meta.get("appPreview", {}).get("static", False):
+            app_path = Path(meta["appPreview"]["file"])
+            if not (rootdir / app_path).is_file():
+                raise FileNotFoundError(
+                    f"{rootdir / app_path} not found (referenced from {compdir / 'index.qmd'})"
+                )
+            yield app_path
+
+
+# components/foo/bar.py => components/static/foo/bar.html
+def infile_to_outfile(infile: Path) -> Path:
+    if infile.parts[0] != "components":
+        raise ValueError(f"Expected components/... path, got {infile}")
+    return Path(infile.parts[0], "static", *infile.parts[1:]).with_suffix(".html")
+
+
+def outfile_to_infile(outfile: Path) -> Path:
+    if (
+        len(outfile.parts) < 3
+        or outfile.parts[0] != "components"
+        or outfile.parts[1] != "static"
+    ):
+        raise ValueError(f"Expected components/static/... path, got {outfile}")
+    return outfile
+
+
+def render_static_preview(appfile: Path, destfile: Path, libdir: Path):
+    appfile = rootdir / appfile
+    destfile = rootdir / destfile
+
+    app_mod = importlib.import_module(
+        str(appfile.relative_to(cwd).with_suffix("")).replace("/", ".")
+    )
+    if "app_ui" not in dir(app_mod):
+        raise RuntimeError(f"app_ui not found in {appfile}")
+    app_ui = app_mod.app_ui
+
+    enrich_app_ui(app_ui)
+
+    destfile.parent.mkdir(parents=True, exist_ok=True)
+
+    app_ui.save_html(destfile, libdir=relative_to(libdir, destfile.parent))
+
+
+def enrich_app_ui(app_ui: Tag):
+    """
+    Need to add Shiny dependency because some inputs/outputs won't look right without
+    being intialized as part of the Shiny binding process.
+    """
+    app_ui.append(shiny.html_dependencies.shiny_deps())
+    app_ui.append(
+        head_content(
+            tags.script(
+                """
+                window.Shiny = window.Shiny || {};
+                window.Shiny.createSocket = function(url) {
+                    // Prevent Shiny from trying to connect to a live server
+                    // by returning a dummy WebSocket object that just no-ops
+                    return {
+                        addEventListener: (event, callback) => {},
+                        send: () => {},
+                        close: () => {},
+                        readyState: 0,
+                        url
+                    };
+                };
+                """
+            )
+        )
+    )
+
+
+def relative_to(path1: Path, path2: Path) -> Path:
+    """
+    Equivalent to `path1.relative_to(path2, walk_up=True)`, except that walk_up=True is
+    only supported from Python 3.12 on and this function works on older versions. If and
+    when we can assume Python 3.12, this function can be removed in favor of the
+    built-in method.
```

Comment:
We're currently building the docs in CI on 3.10, but I don't see why we couldn't bump this to 3.12.

https://github.com/posit-dev/py-shiny-site/blob/c8774a4735f875f08585011f14700cf75db21066/.github/workflows/deploy-docs.yml#L13-L14

We should also probably have a `.python-version` in this repo that matches CI anyway.

So I vote we move up to 3.12. Since that would probably happen in a separate PR, it'd be good to leave a `TODO` comment here to make it easier to find.

Reply:
Oh, great. I had no idea where that was.